index.d.ts

```diff
@@ -191,19 +194,18 @@ declare module "hyperapp" {
   // TypeScript's CSS property definitions. The trade-off doesn't
   // seem worth it given the chances of using such properties.
   // However, you can use type casting if you want to them.
-  type StyleProp = {
+  type StyleProp = IndexableByKey & {
     [K in keyof CSSStyleDeclaration]?: CSSStyleDeclaration[K] | null
-  } & // Since strings are indexable we can avoid them by preventing indexing.
-  { [_: number]: never }
+  }
 
   // Event handlers are implemented using actions.
-  type EventActions<S> = {
-    [K in keyof EventsMap]?: Action<S, EventsMap[K]> | ActionWithPayload<S, any>
+  type EventActions<S, C> = {
+    [K in keyof EventsMap]?: Action<S, EventsMap[K]> | ActionWithPayload<S, C>
```

Comment:
Are we sure this will work for more events on a single node? I mean for something like this (maybe contrived but still, the principle ;) ):

```ts
const AddFoo:Action<State, number> = (state, amount) => ({...state, foo: state.foo + amount})
const SetName:Action<State, string> = (state, name) => ({...state, name})
//...
h('div', {
   onmousedown: [AddFoo, 32],
   onmouseup: [SetName, 'unknown']
})
```

I don't see how `C` could be properly inferred in that case.

For the record, the way that I (in my own types-attempt) handled making sure that payloads in ActionWithPayloads matches their action, is with a "validator type" that I intersect with in the `h` call. Also not 100% perfect but might be interesting to consider: https://github.com/zaceno/zx-typed-todomvc-hyperapp/blob/ed64b9d2ea5e61a2c7b1677b1639f0914fd73509/typings/hyperapp.d.ts#L64

Reply:
Ah, of course, haha. Seems so obvious in hindsight. Yeah this case fails currently.

Reply:
If I understand your types correctly, you get this to not error out by basically not checking the payload: https://github.com/zaceno/zx-typed-todomvc-hyperapp/blob/ed64b9d2ea5e61a2c7b1677b1639f0914fd73509/typings/hyperapp.d.ts#L89

That was my original answer, too. If worse comes to worst then we'll go with that.

However, I'll try a little more to see if there's a more type-safe way to do this that works in the general case.

Reply:
I think you misread that. Notice I'm using `extends` and `infer` keywords! I'm saying "if `X` looks like an ActionWithPayload (while inferring what `S` and `P` actually should be) then X must be assignable to `readonly [Action<S, P>, P]`.

And I do that process for each prop individually. So the problem with this approach is not that it allows too much. On the contrary, the problem is there are edge cases that should be allowed which are not. Like for instance:

```ts

h('div', {
  notAnEventProp: [AddFoo, 'what?! a string?!']
  })
  ```
  
  so my approach would forbid that in the same way it would forbid it if we were assigning the action-with-payload to `onclick`. But since `notAnEventProp` is ... well, not an event prop, it should be allowed to be anything it wants. 
  
  This is such a weird edge case I don't consider it a huge problem. And besides this sort of case my approach works well.

Reply:
And now that I think about it, I could probably amend this problem too. Just not worth it imo.

Reply:
If there is another solution that solves type checking actions with custom payloads it would blow my mind. I think I spent over a month trying to work out this solution so I'm actually kind of proud of it 😅

Reply:
And I had a lot of help from the kind folks on the typescript gitter as well (credit where credit is due)

Reply:
@zaceno You're right, I read too quickly.  I'm exploring more on this with ideas borrowed from your attempt.

Reply:
Just updated the PR with a hopefully complete solution.